### PR TITLE
Add wear to contact action

### PR DESCRIPTION
**Pull request by recuero**
## Reason
 - Make wear consideration in contact user-friendly.
 - Avoid looping over (acting on) secondaries with no corresponding mortar segments 

## Design
No changes.

## Impact
Ease of use.

Refs. #19671

### CODE REVIEW

File: framework/src/auxkernels/MortarNodalAuxKernel.C [hunk 1]
```diff
@@ -78,17 +78,30 @@ MortarNodalAuxKernelTempl<ComputeValueType>::compute()
...
+  for (size_t i = 0; i < its.size(); i++)
+  {
+    mooseAssert(its[i]->first->active(),
+                "We loop over active elements when building the mortar segment mesh, so we golly "
+                "well hope this is active.");
+
+    if (!its[i]->second.empty())
+      // This is local and the mortar segment set isn't empty, so include
+      iterators.push_back(its[i]);
+  }
```
Comment (lindsayad) [inline framework/src/auxkernels/MortarNodalAuxKernel.C:95]: I don't understand this. Do we have cases where there are no mortar segments associated with a secondary? That seems like a problem in the mortar mesh generation or I guess in the generation of this data.

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: framework/src/auxkernels/MortarNodalAuxKernel.C (hunk 1 above)
Comment (recuero) [reply framework/src/auxkernels/MortarNodalAuxKernel.C:95]: As the code stands right now, we have secondary elements with no corresponding mortar segment quadrature points (I assume just corresponding mortar segments). Then we make calls in `MortarUtils.h` like this:

``` if (assembly.needDual() )
      assembly.reinitDual(secondary_face_elem, secondary_xi_pts, JxW);
```
      
which require having some mortar segments associated with secondary elements, otherwise we get exceptions from libMesh. 

> Do we have cases where there are no mortar segments associated with a secondary?

I guess in a general case we can have cases where there are no mortar segment associated with a secondary element (say a circle in contact with a small plane, where the circle is secondary surface).

Comment (recuero) [reply framework/src/auxkernels/MortarNodalAuxKernel.C:95]: There is probably a better way to deal with this.

Comment (lindsayad) [reply framework/src/auxkernels/MortarNodalAuxKernel.C:95]: > I guess in a general case we can have cases where there are no mortar segment associated with a secondary element (say a circle in contact with a small plane, where the circle is secondary surface).

In the early days we would have had mortar segments from these, they just wouldn't have any bisections (loose use of bisection here) by inverse projection of primary nodes onto the secondary face element, e.g. the mortar segment elements would have corresponded exactly to the secondary face elements. However, @jbadger95 may have removed these IIRC, e.g. we only have mortar segment elements for cases in which the secondary projects onto the primary. Perhaps this is why `second` can be empty now?

Comment (jbadger95) [reply framework/src/auxkernels/MortarNodalAuxKernel.C:95]: As far as I remember this is correct @lindsayad. There used to be a `has_primary` flag in mortar segments but because there was a lot of confusion around edge dropping we changed it so that only mortar segments with a primary element are added to the mortar segment mesh.

Comment (lindsayad) [reply framework/src/auxkernels/MortarNodalAuxKernel.C:95]: @recuero would you want to check and see whether you can eliminate the presence of "empty" secondaries from the data returned by `secondariesToMortarSegments` so that we don't have to do manual manipulation in user code (like in `MortarNodalAuxKernel`)? If you don't have time, I can look into it

Comment (recuero) [reply framework/src/auxkernels/MortarNodalAuxKernel.C:95]: Yeah, that sounds like the way to go. I can look into that. Give me a bit of time.

**Comment by lindsayad** [comment]
@recuero feel free to review my changes

File: framework/src/constraints/AutomaticMortarGeneration.C [hunk 2]
```diff
@@ -2230,24 +2241,21 @@ AutomaticMortarGeneration::secondariesToMortarSegments(const Node & node) const
...
 
   for (const auto i : index_range(secondary_elems))
   {
     auto * const secondary_elem = secondary_elems[i];
     auto msm_it = _secondary_elems_to_mortar_segments.find(secondary_elem);
     if (msm_it == _secondary_elems_to_mortar_segments.end())
-      // We may have removed this element key from this map
-      continue;
-
+      mooseError("end of mortar segment map in associatedMortarSegments");
```
Comment (recuero) [inline framework/src/constraints/AutomaticMortarGeneration.C:2251]: Each secondary element in the maps must now always have a corresponding mortar segment by mesh construction (my understanding).

Comment (recuero) [reply framework/src/constraints/AutomaticMortarGeneration.C:2251]: We are removing the secondaries from the nodal map if the msm is empty: Is that general enough to enforce this error?

File: framework/src/constraints/AutomaticMortarGeneration.C [hunk 3]
```diff
@@ -851,15 +857,19 @@ AutomaticMortarGeneration::buildMortarSegmentMesh3d()
...
-      auto & secondary_to_msm_element_set =
-          _secondary_elems_to_mortar_segments[secondary_side_elem];
 
       const Real secondary_volume = secondary_side_elem->volume();
 
       // If this Elem is not in the current secondary subdomain, go on to the next one.
       if (secondary_side_elem->subdomain_id() != secondary_subd_id)
         continue;
 
+      auto [secondary_elem_to_msm_map_it, insertion_happened] =
```
[the remark this answers is not in the document]
Comment (lindsayad) [reply framework/src/constraints/AutomaticMortarGeneration.C:878]: This is a structured binding to the returned pair. You have to bind a variable to each member of the returned pair; you can't ignore one of the returned items like you can in python with `_`.

File: framework/src/constraints/AutomaticMortarGeneration.C (hunk 2 above)
Comment (lindsayad) [reply framework/src/constraints/AutomaticMortarGeneration.C:2251]: Yes it should be. We should not have any secondaries without msms either in the node-to-secondary map or the secondary-to-msm map.

Comment (recuero) [reply framework/src/constraints/AutomaticMortarGeneration.C:2251]: We may be missing something. I created a typical bouncing block example with a secondary sideset that wraps around its surface and came across that error.

Comment (lindsayad) [reply framework/src/constraints/AutomaticMortarGeneration.C:2251]: Haha look at you coming up with more tricky tests! This is good. If you can, please push up that test, and I will take a look

File: framework/src/auxkernels/MortarNodalAuxKernel.C [hunk 4]
```diff
@@ -78,8 +80,8 @@ MortarNodalAuxKernelTempl<ComputeValueType>::compute()
   ComputeValueType value(0);
   Real total_volume = 0;
 
-  const auto & its = amg().secondariesToMortarSegments(*_current_node);
-  std::array<MortarNodalAuxKernelTempl<ComputeValueType> *, 1> consumers = {{this}};
+  const std::vector<AutomaticMortarGeneration::MortarFilterIter> & its =
+      amg().secondariesToMortarSegments(*_current_node);
```
[the remark this answers is not in the document]
Comment (recuero) [reply framework/src/auxkernels/MortarNodalAuxKernel.C]: I think `auto` tends to make the code less readable, but I am swimming against the tide. Reverted it in the latest commit.

File: framework/src/constraints/AutomaticMortarGeneration.C (hunk 2 above)
Comment (recuero) [reply framework/src/constraints/AutomaticMortarGeneration.C:2251]: Just did!

File: framework/include/constraints/AutomaticMortarGeneration.h [hunk 5]
```diff
@@ -503,6 +505,12 @@ class AutomaticMortarGeneration : public ConsoleStreamInterface
   /// Flag to enable regressed treatment of edge dropping where all LM DoFs on edge dropping element
   /// are strongly set to 0.
   bool _correct_edge_dropping;
```
Comment (lindsayad) [inline framework/include/constraints/AutomaticMortarGeneration.h:507]:
```suggestion
  const bool _correct_edge_dropping;
```

File: framework/include/constraints/AutomaticMortarGeneration.h [hunk 6]
```diff
@@ -503,6 +505,12 @@ class AutomaticMortarGeneration : public ConsoleStreamInterface
...
+
+  /// Parameter to control which angle (in degrees) is admissible for the creation of mortar segments.
+  /// If set to a value close to zero, very oblique projections are allowed, which can result in mortar
+  /// segments solving physics not meaningfully and overprojection of primary nodes onto the mortar
+  /// segment mesh in extreme cases. This parameter is mostly intended for mortar mesh debugging purposes in 2D.
+  Real _minimum_projection_angle;
```
Comment (lindsayad) [inline framework/include/constraints/AutomaticMortarGeneration.h:513]:
```suggestion
  const Real _minimum_projection_angle;
```

File: framework/src/constraints/AutomaticMortarGeneration.C [hunk 7]
```diff
@@ -706,6 +707,18 @@ AutomaticMortarGeneration::buildMortarSegmentMesh()
...
+
+  // Deleting elements may produce isolated nodes.
+  // Loops for identifying and removing such nodes from mortar segment mesh.
+  for (const auto & element : _mortar_segment_mesh->element_ptr_range())
+    for (auto & n : element->node_ref_range())
+      msm_connected_nodes.insert(&n);
+
+  for (const auto & node : _mortar_segment_mesh->node_ptr_range())
+    if (!msm_connected_nodes.count(node))
+      _mortar_segment_mesh->delete_node(node);
```
Comment (lindsayad) [inline framework/src/constraints/AutomaticMortarGeneration.C:720]: Do these orphaned nodes have a practical impact on the computation? Or does this make a debug mesh or nodal geometry look weird if they're left in? Just curious if there's a motivation for this other than making the mesh "logically clean"

File: framework/src/constraints/AutomaticMortarGeneration.C [hunk 8]
```diff
@@ -1572,17 +1591,36 @@ AutomaticMortarGeneration::computeNodalGeometry()
...
+    nodal_tangent_one(0) = 0;
+    nodal_tangent_one(1) = 1;
+    nodal_tangent_one(2) = 0;
+
+    nodal_tangent_two(0) = 0;
+    nodal_tangent_two(1) = 0;
+    nodal_tangent_two(2) = -1;
+
+    return;
+  }
```
Comment (lindsayad) [inline framework/src/constraints/AutomaticMortarGeneration.C:1623]: I assume there was a methodology to this choice? If it's not too complicated, you might consider adding a comment regarding it

**Review by lindsayad — CHANGES_REQUESTED** [review]
Getting to the finish line I think!

File: framework/src/constraints/AutomaticMortarGeneration.C (hunk 8 above)
Comment (recuero) [reply framework/src/constraints/AutomaticMortarGeneration.C:1623]: Yeah, the equations yield these vectors as we approach the singularity. Will add a comment to that respect.

**Comment by lindsayad** [comment]
I think it might be worth adding a comment, maybe in the place where we finally delete the msm from the mesh (in 2D), that we may be creating nodes with only one element neighbor where before there were two, but that the nodal normal used in computations will reflect the two-neighbor geometry

File: framework/src/constraints/AutomaticMortarGeneration.C (hunk 7 above)
Comment (recuero) [reply framework/src/constraints/AutomaticMortarGeneration.C:720]: I don't think they have an impact on the computation; it doesn't for the case I tried out. But those nodes still reside in the `mortar_segment_mesh`, which makes the exodus file (object) different from the case in which we do not wrap the sidesets/lowerds around the block. My main motivation was to ensure the `mortar_segment_mesh` is as close as possible to the reference situation with lower-d only on the contact patch, otherwise, for the bouncing block, e.g., the mesh has double the number of nodes compared to the reference case.

File: framework/src/constraints/AutomaticMortarGeneration.C [hunk 9]
```diff
@@ -696,7 +691,13 @@ AutomaticMortarGeneration::buildMortarSegmentMesh()
...
     {
       // Erase from secondary to msms map
-      libmesh_map_find(_secondary_elems_to_mortar_segments, msinfo.secondary_elem).erase(msm_elem);
+      auto it = _secondary_elems_to_mortar_segments.find(msinfo.secondary_elem);
+      mooseAssert(it != _secondary_elems_to_mortar_segments.end(),
+                  "We should have found the element");
+      auto & msm_set = it->second;
+      msm_set.erase(msm_elem);
+      if (msm_set.empty())
+        _secondary_elems_to_mortar_segments.erase(it);
```
Comment (lindsayad) [inline framework/src/constraints/AutomaticMortarGeneration.C:700]: Actually I think this would be the most appropriate for the nodal normal comment, since it is the nodes with secondary element neighbors (not mortar segment mesh neighbors) that determine the nodal normal field

File: framework/src/constraints/AutomaticMortarGeneration.C (hunk 7 above)
Comment (lindsayad) [reply framework/src/constraints/AutomaticMortarGeneration.C:720]: Alright we can keep this then. If it shows up in profiling ever, we can consider removing it

File: framework/src/constraints/AutomaticMortarGeneration.C (hunk 8 above)
Comment (recuero) [reply framework/src/constraints/AutomaticMortarGeneration.C:1623]: This is an example of how the fields look like on a spherical surface for a node that doesn't go in the avoid-singularity if statement:

image

**Review by lindsayad — DISMISSED** [review]
Maybe we should have @dschwen or someone else do a final run-through since I have a lot of commits in this PR?

File: framework/src/auxkernels/MortarNodalAuxKernel.C [hunk 10]
```diff
@@ -46,7 +48,9 @@ MortarNodalAuxKernelTempl<ComputeValueType>::MortarNodalAuxKernelTempl(
     MortarExecutorInterface(
         *this->template getCheckedPointerParam<FEProblemBase *>("_fe_problem_base")),
     MortarConsumerInterface(this),
-    _displaced(this->template getParam<bool>("use_displaced_mesh")),
+    _displaced(parameters.isParamSetByUser("use_displaced_mesh")
+                   ? this->template getParam<bool>("use_displaced_mesh")
+                   : true),
```
Comment (dschwen) [inline framework/src/auxkernels/MortarNodalAuxKernel.C:53]: do 

```
params.set<bool>("use_displaced_mesh") = true; 
```
in validParams instead.

File: framework/src/auxkernels/MortarNodalAuxKernel.C [hunk 11]
```diff
@@ -104,6 +109,12 @@ MortarNodalAuxKernelTempl<ComputeValueType>::compute()
   // We have to reinit the node for this variable in order to get the dof index set for the node
   _var.reinitNode();
 
+  // If the node doesn't have corresponding mortar segments, force the value assigned in this step
+  // to be zero. This can be useful when nodes initially do not project but will project at a
+  // different stage of the simulation
+  if (total_volume == 0)
+    total_volume = std::numeric_limits<Real>::max();
```
Comment (dschwen) [inline framework/src/auxkernels/MortarNodalAuxKernel.C:116]: why not rather do 

```
  if (total_volume == 0)
    value = 0;
  else
    value /= total_volume;
```
dividing by max real later is a bit of a detour IMO

Comment (dschwen) [reply framework/src/auxkernels/MortarNodalAuxKernel.C:116]: maybe a fuzzy equal?

Comment (recuero) [reply framework/src/auxkernels/MortarNodalAuxKernel.C:116]: We are setting value/total_volume below

File: modules/contact/test/tests/3d-mortar-contact/tests [hunk 12]
```diff
@@ -59,19 +59,6 @@
...
-    min_ad_size = 100
-    exodiff = 'nodal_geometry_only.e'
-    mesh_mode = 'REPLICATED'
-    cli_args = 'Executioner/end_time=0.5 Executioner/dtmin=0.5 Constraints/normal_lm/debug_mesh=true '
-               'Constraints/normal_x/debug_mesh=true Constraints/normal_y/debug_mesh=true '
-               'Constraints/normal_z/debug_mesh=true Outputs/exodus=false Problem/solve=false'
-    requirement = 'The system shall solve a 3D frictionless bouncing block problem with mortar '
-                  'constraint and output the normal and tangent vector fields for debugging purposes'
-    max_parallel = 1
-  []
```
Comment (dschwen) [inline modules/contact/test/tests/3d-mortar-contact/tests:74]: why is this removed?

File: framework/src/constraints/AutomaticMortarGeneration.C [hunk 13]
```diff
@@ -1572,17 +1597,39 @@ AutomaticMortarGeneration::computeNodalGeometry()
...
+  const Real nx = nodal_normal(0);
+  const Real ny = nodal_normal(1);
+  const Real nz = nodal_normal(2);
+
   // See Lopes DS, Silva MT, Ambrosio JA. Tangent vectors to a 3-D surface normal: A geometric tool
   // to find orthogonal vectors based on the Householder transformation. Computer-Aided Design. 2013
   // Mar 1;45(3):683-94.
-  const Point h_vector(
-      std::max(unit_normal(0) - 1.0, unit_normal(0) + 1.0), unit_normal(1), unit_normal(2));
+  const Point h_vector(std::max(nx - 1.0, nx + 1.0), ny, nz);
```
Comment (dschwen) [inline framework/src/constraints/AutomaticMortarGeneration.C:1616]: How is `std::max(nx - 1.0, nx + 1.0)` ever not `nx + 1.0` ?

File: framework/src/constraints/AutomaticMortarGeneration.C [hunk 14]
```diff
@@ -1727,21 +1775,33 @@ AutomaticMortarGeneration::projectSecondaryNodesSinglePair(
...
+            {
+              Real dxi2 = -F.value() / F.derivatives();
 
-            xi2_dn += dxi2;
+              xi2_dn += dxi2;
+            }
+            else
+              // It's possible that the nodal normal is completely orthogonal to the primary
+              // surface, in which case the derivative is 0. We know in this case that the
+              // projection should be a failure
```
Comment (dschwen) [inline framework/src/constraints/AutomaticMortarGeneration.C:1793]: I don't understand this comment. Wouldn't that be a fairly common case in parallel plate contact?

**Review by dschwen — COMMENTED** [review]
See comments

File: framework/src/constraints/AutomaticMortarGeneration.C (hunk 14 above)
Comment (recuero) [reply framework/src/constraints/AutomaticMortarGeneration.C:1793]: What's the implication that this case can happen commonly? I don't fully understand your comment.

File: modules/contact/test/tests/mortar_aux_kernels/block-dynamics-aux-fretting-wear-test-action.i [hunk 15]
```diff
@@ -0,0 +1,283 @@
...
+  [disp_y]
+    block = 2
+    variable = disp_y
+    value = '${fparse starting_point + offset}'
+    type = ConstantIC
+  []
+[]
+
+[Kernels]
+  [DynamicTensorMechanics]
```
[the remark this answers is not in the document]
Comment (recuero) [reply modules/contact/test/tests/mortar_aux_kernels/block-dynamics-aux-fretting-wear-test-action.i:39]: It works. I'll be using it in an upcoming commit, but needed to regold because found differences on the order of 1.0e-3 which can be attributed to the execution of the auxiliary kernels for velocity and acceleration.

**Comment by dschwen** [comment]
`The requested mortar interface AutomaticMortarGeneration object does not yet exist!` something broke. :-(

File: framework/src/constraints/AutomaticMortarGeneration.C (hunk 14 above)
Comment (lindsayad) [reply framework/src/constraints/AutomaticMortarGeneration.C:1793]: @dschwen I changed "orthogonal to the primary surface" to "orthogonal to the primary surface normal". Hopefully that clarifies things?

Comment (lindsayad) [reply framework/src/constraints/AutomaticMortarGeneration.C]: and for good measure "nodal normal" to "secondary surface nodal normal"

**Comment by recuero** [comment]
I cannot find a way to account for the existence of `displaced_mesh` in the relationship manager parameter of `MortarConsumerInterface`. I believe we'd need that to make the call to `getMortarInterface` from `AugmentSparsityOnInterface` have the correct value for `_use_displaced_mesh`, as this parameter is changed in `FEProblemBase` when creating kernels and auxiliary kernels. Maybe @lindsayad has a better idea.

**Comment by lindsayad** [comment]
Why did this issue crop up? (I believe there is an issue, but how are we seeing it now and not before?)

**Comment by lindsayad** [comment]
Never mind, I see the most recent commit explains it. That is unfortunate but I don't have a good/easy solution for that 😦

**Comment by recuero** [comment]
It's arising now because we are trying to set defaults in use_displaced_mesh for mortar objects that can act on displaced and undisplaced meshes.

**Comment by lindsayad** [comment]
it would be nice to get this merged. @dschwen could you check and see if the conversations are resolved?

**Review by lindsayad — APPROVED** [review]
I believe all review comments have been addressed and there is mortar work that may depend on this
